### PR TITLE
[FSDP][state_dict][1/N] Moving state_dict logic to pre_state_dict_hook

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* #88635
* #88481
* __->__ #87900

This is one step toward the ultimate goal: remove the overwritten state_dict in FSDP. All the logic should be either in `pre_state_dict_hook` or `post_state_dict_hook`. 

Since current `nn.Module` does not support `pre_state_dict_hook`, this PR mimic `pre_state_dict_hook` by calling the pre hook inside post the hook, effectively ditching all the work done by `nn.Module.state_dict`. Once `pre_state_dict_hook` is supported by `nn.Module`, these pre hook calls can be moved out from the post hooks and be registered to `nn.Module.pre_state_dict_hook`.

The major issue of this temporary solution is that `post_state_dict_hook` is called from the leaf node to the root node. This makes the `module._lazy_init()` invalid as FSDP assumes `_lazy_init()` to be called from the root. As a result, `FSDP.state_dict` currently contains only one logic -- calling `module._lazy_init()`.

### CODE REVIEW

**Review by zhaojuanmao — COMMENTED** [review]
awesome! overall look great to me, will let Rohan or Andrew accept

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 1]
```diff
@@ -345,7 +345,7 @@ def test_state_dict_rank0_offload_save_load_flow(self, use_orig_params: bool):
     )
     @parametrize("fp16", [True, False])
     @parametrize("state_dict_rank0_and_offload", [True, False])
-    @parametrize("use_orig_params", [False, True])
+    @parametrize("use_orig_params", [True, False])
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_state_dict.py:348]: nit: why do we need to change this order?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 2]
```diff
@@ -26,48 +27,217 @@
...
+    module._clear_grads_if_needed()
+
+
+def _common_summon_pre_state_dict_hook(
+    module,
+    offload_to_cpu: bool,
+    rank0_only: bool,
+) -> None:
+    """
+    Performs the pre-state_dict tasks sharded by all state_dict types that require
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:87]: nit: "sharded" -> shared?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 3]
```diff
@@ -26,48 +27,217 @@
...
+        return state_dict
 
     # Loop only the parameters saved in this instance's wrapped module to
     # avoid processing buffers.
     for fqn, param_name, module_name in module._param_fqns:
+        # TODO: remove the parameter retrieval. See ``_full_pre_state_dict_hook``.
+        param = functools.reduce(getattr, fqn.split("."), module.module)
         fqn = f"{prefix}{fqn}"
+        if no_fsdp_return:
+            state_dict.pop(fqn)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:160]: is this to remove FSDP managed parameters from the state_dict if we are rank0_only and not rank 0?

Comment (rohan-varma) [reply torch/distributed/fsdp/_state_dict_utils.py:160]: And would we still need this once we have the proper support for pre_state_dict hook in nn.Module?

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 4]
```diff
@@ -26,48 +27,217 @@
...
+            continue
+        state_dict[fqn] = param
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_state_dict_utils.py:162]: is this where we're actually populating the state_dict? i.e. we're not relying on anything that the nn.Module state_dict implementation does anymore?

**Review by rohan-varma — APPROVED** [review]
LGTM overall, but agree this is a somewhat hacky solution because we're calling the state_dict, then in post-hook processing essentially implementing our own version of state_dict by running a pre hook, discarding super().state_dict results, and then doing post-hook. 

Adding nn.Module pre state_dict hook immediately is probably better solution, but might take ~2-3 weeks. If that timeline doesn't work then we can go ahead and land this change.

File: test/distributed/fsdp/test_fsdp_state_dict.py (hunk 1 above)
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_state_dict.py:348]: uh, this is just to make this line be consistent with the previous lines.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 3 above)
Comment (fegin) [reply torch/distributed/fsdp/_state_dict_utils.py:160]: 1. Yes
2. IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if `rank_only and rank !=0`. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside `state_dict` on other ranks.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 4 above)
Comment (fegin) [reply torch/distributed/fsdp/_state_dict_utils.py:162]: Yes.

File: torch/distributed/fsdp/_state_dict_utils.py (hunk 3 above)
Comment (awgu) [reply torch/distributed/fsdp/_state_dict_utils.py:160]: > IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if rank_only and rank !=0. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside state_dict on other ranks.

Good call!

File: torch/distributed/fsdp/_state_dict_utils.py [hunk 5]
```diff
@@ -31,49 +38,218 @@
...
+    prefix: str,
+) -> None:
+    """Performs the pre-state_dict tasks shared by all state_dict types."""
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+    _lazy_init(module, module)
+    # TODO: change to this call after pre_state_dict_hook is in `nn.Module`.
+    # if module.is_root:
+    #    _clear_grads_if_needed(module._fsdp_handles(module))
+    if module._has_params:
```
Comment (awgu) [inline torch/distributed/fsdp/_state_dict_utils.py:92]: Sorry, let me clarify the `_clear_grads_if_needed()` call.

The idea is that we want to free the gradient memory as early as possible, so, given that this `_clear_grads_if_needed()` incurs minimal overhead (looping over gradients and checking if all are `None`), I inserted this call at every FSDP callsite.

In that way, we should prefer to keep it as is since the root can help free all possible gradient memory, which may be earlier than deferring to the submodule.

Comment (fegin) [reply torch/distributed/fsdp/_state_dict_utils.py:92]: Right now, the pre_hook is not an actual pre_state_dict_hook. So it will be called from the bottom-up. So the gradient won't be freed in time for a leaf module before the leaf module's post_state_dict_hook is called. Once the pre_state_dict_hook is actually implemented in `nn.Module`, I'll change it back.
